compiler/src/dotty/tools/dotc/cc/CCState.scala

```diff
@@ -0,0 +1,165 @@
+
+/** Capture checking state, which is known to other capture checking components */
+class CCState:
+  import CCState.*
+
+  // ------ Error diagnostics -----------------------------
+
+  /** Error reprting notes produces since the last call to `test` */
+  var notes: List[ErrorNote] = Nil
+
+  def addNote(note: ErrorNote): Unit =
+    if !notes.exists(_.getClass == note.getClass) then
+      notes = note :: notes
+
+  def test(op: => CompareResult): CompareResult =
+    val saved = notes
+    notes = Nil
+    try op match
+      case res: CompareFailure => res.withNotes(notes)
+      case res => res
+    finally notes = saved
+
+  def testOK(op: => Boolean): CompareResult =
+    test(if op then CompareResult.OK else CompareResult.Fail(Nil))
+
+  /** Warnings relating to upper approximations of capture sets with
+   *  existentially bound variables.
+   */
+  val approxWarnings: mutable.ListBuffer[Message] = mutable.ListBuffer()
+
+  // ------ Level handling ---------------------------
+
+  private var curLevel: Level = outermostLevel
+
+  /** The level of the current environment. Levels start at 0 and increase for
+   *  each nested function or class. -1 means the level is undefined.
+   */
+  def currentLevel(using Context): Level = curLevel
+
+  /** Perform `op` in the next inner level */
+  inline def inNestedLevel[T](inline op: T)(using Context): T =
+    val saved = curLevel
+    curLevel = curLevel.nextInner
+    try op finally curLevel = saved
+
+  /** Perform `op` in the next inner level unless `p` holds. */
+  inline def inNestedLevelUnless[T](inline p: Boolean)(inline op: T)(using Context): T =
+    val saved = curLevel
+    if !p then curLevel = curLevel.nextInner
+    try op finally curLevel = saved
+
+  /** A map recording the level of a symbol */
+  private val mySymLevel: mutable.Map[Symbol, Level] = mutable.Map()
+
+  def symLevel(sym: Symbol): Level = mySymLevel.getOrElse(sym, undefinedLevel)
+
+  def recordLevel(sym: Symbol)(using Context): Unit = mySymLevel(sym) = curLevel
+
+  // ------ BiTypeMap adjustment -----------------------
+
+  private var myMapFutureElems = true
+
+  /** When mapping a capture set with a BiTypeMap, should we create a BiMapped set
+   *  so that future elements can also be mapped, and elements added to the BiMapped
+   *  are back-propagated? Turned off when creating capture set variables for the
+   *  first time, since we then do not want to change the binder to the original type
+   *  without capture sets when back propagating. Error case where this shows:
+   *  pos-customargs/captures/lists.scala, method m2c.
+   */
+  def mapFutureElems(using Context) = myMapFutureElems
+
+  /** Don't map future elements in this `op` */
+  inline def withoutMappedFutureElems[T](op: => T)(using Context): T =
+    val saved = mapFutureElems
+    myMapFutureElems = false
+    try op finally myMapFutureElems = saved
+
+  // ------ Iteration count of capture checking run
+
+  private var iterCount = 1
```

Comment:
Just wondering, do you know the highest number the counter goes when running the tests?

Reply:
I think there should be at most one extra iteration. I believe that's the case always, but I am not sure about that.